bayesian_protein/surrogates.py

```diff
@@ -525,6 +761,178 @@ def forward(self, X: np.ndarray, smiles) -> Tuple[np.ndarray, np.ndarray]:
+        self.models = []
+        for i in range(5):
+            m = nn.Sequential(
+                    nn.Linear(input_size, 256),
+                    nn.GELU(),
+                    #nn.Dropout(p=0.2),
+                    nn.Linear(256, 24),
+                    nn.GELU(),
+                    #nn.Dropout(p=0.2),
+                    nn.Linear(24, 1),
+                ).to(self.device)
+            self.models.append(m)
+        self.known_input_size = input_size
+
+    def _restore(self):
+        if self.models is None:
+            raise ValueError(
+                "Can only be called once 'update' has been called at least once."
+            )
+
+        self._build_model(self.known_input_size)
+
+    def update(self, x: np.ndarray, y: List[float], smiles: List[str]):
+        self._build_model(x.shape[1])
+
+        super().update(x, y, smiles)
+
+    def _train_loop(self, model, loader: DataLoader, optimizer: torch.optim.Optimizer, lr_scheduler):
+        if len(loader) <= 0:
+            return 99.9
+        avg = 0
+        model.train()
+        for i, (x, y) in enumerate(loader, start=1):
+            x, y = x.to(self.device), y.squeeze().to(self.device)
+            optimizer.zero_grad()
+            output = model.forward(x).squeeze()
+            loss = nn.functional.huber_loss(output, y)
+            loss.backward()
+            optimizer.step()
+            lr_scheduler.step()
+
+            # Update running average of loss
+            avg += 1 / i * (loss.mean().item() - avg)
+
+        return avg
+
+    @torch.no_grad()
+    def _val_loop(self, model, loader: DataLoader):
+        if len(loader) <= 0:
+            return 99.9
+        avg = 0
+        model.eval()
+        for i, (x, y) in enumerate(loader, start=1):
+            x, y = x.to(self.device), y.squeeze().to(self.device)
+            output = model.forward(x).squeeze()
+            loss = nn.functional.huber_loss(output, y)
+            avg += 1 / i * (loss.mean().item() - avg)
+
+        return avg
+
+    def _fit(self, X: np.ndarray, y: np.ndarray):
+        if self.models is None:
+            raise ValueError(
+                "Can only be called once 'update' has been called at least once."
+            )
+        if self.retrain:
+            self._restore()
+
+        # Setup data
+        tensor_x = torch.as_tensor(X, dtype=torch.float).cuda()
+        tensor_y = torch.as_tensor(y, dtype=torch.float).cuda()
+        ds = TensorDataset(tensor_x, tensor_y)
+        b1, b2, b3, b4, b5 = random_split(ds, lengths=[0.2, 0.2, 0.2, 0.2, 0.2])
+        blocks = [b1,b2,b3,b4,b5]
+        for i in range(len(self.models)):
+            train_ds_block = []
+            for q in range(len(self.models)):
+                if q==i:
+                    continue
+                train_ds_block.append(blocks[q])
+            train_ds = torch.utils.data.ConcatDataset(train_ds_block)
+            val_ds = blocks[i]
+            if len(train_ds) <= 0:
```

Comment:
Can this happen?

Reply:
Should be fixed with the latest [commit](https://github.com/uds-lsv/chemical-lm-active-learning/pull/2/commits/53edd12ef9b7b3d199a02bbad7aa1e1b43b76f15)